Dockerfile

```diff
@@ -5,14 +5,9 @@ FROM golang:1.12.7-alpine AS builder
 
 RUN apk add --update --no-cache ca-certificates make git curl mercurial bzr
 
-WORKDIR /go/src/github.com/uber/cadence
-
-RUN curl https://raw.githubusercontent.com/golang/dep/master/install.sh | INSTALL_DIRECTORY=/usr/local/bin sh
-COPY Gopkg.* ./
```

Comment:
This should be replaced with go.* copied into the image and `go mod download` ran in order to build an image cache.

Here is a snippet from our [Dockerfile](https://github.com/banzaicloud/pipeline/blob/master/Dockerfile):

```
ENV GOFLAGS="-mod=readonly"
ARG GOPROXY

COPY go.* .
RUN go mod download
```

The readonly flag makes sure that no new dependencies are installed, thus making sure that there are no surprises around dependencies.

The GOPROXY arg allows using a go modules proxy. It's optional, but allows a developers to build an image with a proxy.

Reply:
thanks for pointing out, the `ENV GOFLAGS="-mod=readonly"` and `ARG GOPROXY` are added.

just 2 questions
1. on line 10 there is a `COPY . .`, which should copy the  `go.mod` and `go.sum`?
2. why `go mod download`? I tested building the docker image locally, and it seems working.

Reply:
In order to utilize the docker build cache, it's a best practice to place things at the beginning of the build that has the least chance to change.

In this case, by downloading dependencies separately first you can spare some time with the build when only the code changes, but not the list of dependencies (which is usually the case).

Otherwise every time something changes, you would have to download the dependencies (which takes time) even if they haven't actually changed.

Although this does not add much in a build environment (if you can't utilize the build cache), it's really helpful when building locally. It doesn't take more time to build the image this way, so this kind of optimization usually make sense.

Reply:
i see